### DISCUSSION TITLE
Inconsistency in search_for Results: Seeking Single fitz.Rect for Entire Phrases in PyMuPDF

### DISCUSSION THREAD

**Discussion by Mazzesy** [pymupdf/PyMuPDF#3046] [Looking for help]
### Description of the bug

I am utilizing the search_for function from the PyMuPDF library to identify the positions of phrases within a PDF document. Specifically, I am working with this document ([Link](https://eur-lex.europa.eu/legal-content/DE/TXT/?uri=CELEX%3A32022R2379)).

When I search for the phrase VERORDNUNG (EU) 2022/2379, the search_for function returns a single fitz.Rect representing the entire phrase. However, when searching for Verordnung (EG) Nr. 617/2008, the function provides separate fitz.Rect instances for each part of the phrase.

My objective is to consistently obtain a single fitz.Rect for the entire phrase. Could someone please explain the reason behind this discrepancy in the function's output?

I open the same question in stack overflow ([Link](https://stackoverflow.com/questions/77820264/inconsistency-in-search-for-results-seeking-single-fitz-rect-for-entire-phrases)).

### How to reproduce the bug

> import fitz
> 
> phrase = "phrase"
> path = "path/to/pdf"
> 
> with fitz.open(path) as doc:
>     page = doc[0]
>     text_instances = page.search_for(phrase)
>     print(text_instances)

Pdf file that produces the inconsistency: [CELEX 32022R2379 DE TXT.pdf](https://github.com/pymupdf/PyMuPDF/files/13942458/CELEX.32022R2379.DE.TXT.pdf)

### PyMuPDF version

1.23.14

### Operating system

MacOS

### Python version

3.9

**Comment by JorjMcKie** [comment]
I haven't looked at all the details yet, But I saw that your issue happens in paragraphs having justified text alignment.
In these cases, MuPDF - depending on the inter-word distances - may even insert linebreaks (!) between words, because it cannot know the intention of these large distances.
You have two alternatives:
1. Set off text extraction flag bits [preserve white spaces](file:///C:/Users/haral/OneDrive/Desktop/doc-new/build/pymupdf/vars.html#TEXT_PRESERVE_WHITESPACE) and [inhibit spaces](file:///C:/Users/haral/OneDrive/Desktop/doc-new/build/pymupdf/vars.html#TEXT_INHIBIT_SPACES). This will convert all white space to spaces and fill extraneous inter-word distances with spaces. This may help - you will need to try things out potentially.
2. When getting several response rectangles that have the same bottom coordinate, then you know the situation and simply join these rectangles.

**Reply by Mazzesy** [reply]
I have tried the second variant and combined the rectangles when they are on the same line. How can I achieve the first variant?

The code I use for number 2 is this:

```
def combine_rectangles_on_same_line(rectangles):
    result = []
    grouped_rectangles = []

    # Sort rectangles by their y-coordinate
    rectangles.sort(key=lambda rect: rect.y0)

    for rect in rectangles:
        # If grouped_rectangles is empty or the current rect is on the same line
        if not grouped_rectangles or is_same_line(grouped_rectangles[-1], rect):
            grouped_rectangles.append(rect)
        else:
            # Combine rectangles on the same line with different text
            combined_rect = combine_rectangles(grouped_rectangles)
            result.append(combined_rect)
            grouped_rectangles = [rect]

    # Add the last group of rectangles
    if grouped_rectangles:
        combined_rect = combine_rectangles(grouped_rectangles)
        result.append(combined_rect)

    return result

def is_same_line(rect1, rect2, tolerance=5):
    # Check if two rectangles are on the same line with a tolerance value
    return abs(rect1.y0 - rect2.y0) < tolerance

def combine_rectangles(rectangles):
    # Combine rectangles into a single rectangle
    combined_rect = rectangles[0]
    for rect in rectangles[1:]:
        combined_rect = fitz.Rect(
            min(combined_rect.x0, rect.x0),
            min(combined_rect.y0, rect.y0),
            max(combined_rect.x1, rect.x1),
            max(combined_rect.y1, rect.y1)
        )

    return combined_rect

with fitz.open(path) as doc:
    page = doc[0]
    text_instances = page.search_for(phrase)
    combined_rectangles = combine_rectangles_on_same_line(text_instances)
```

**Reply by zamrara** [reply]
Much shorter version for number 2 that worked for me was:
```python
import fitz 

def combine_rectangles(rectangles):
    combined_rect = rectangles[0]
    for rect in rectangles[1:]:
        combined_rect.include_rect(rect)
    return combined_rect

with fitz.open(path) as doc:
    page = doc[0]
    text_instances = page.search_for(phrase)
    combined_rectangles = combine_rectangles(text_instances)

```

**Comment by JorjMcKie** [comment]
I will move this post to Discussions for further communication.
